libensemble/utils/runners.py

```diff
@@ -84,3 +97,117 @@ def _result(self, calc_in: npt.NDArray, persis_info: dict, libE_info: dict) -> (
     def shutdown(self) -> None:
         if self.thread_handle is not None:
             self.thread_handle.terminate()
+
+
+class AskTellGenRunner(Runner):
+    """Interact with ask/tell generator. Base class initialized for third-party generators."""
+
+    def __init__(self, specs):
+        super().__init__(specs)
+        self.gen = specs.get("generator")
+        self.initial_batch = getattr(self.gen, "initial_batch_size", 0)
+        self.batch = getattr(self.gen, "batch_size", 0)
+
+    def _to_array(self, x: list) -> npt.NDArray:
+        """fast-cast list-of-dicts to NumPy array"""
+        if isinstance(x, list) and len(x) and isinstance(x[0], dict):
+            arr = np.zeros(len(x), dtype=self.specs["out"])
+            for i in range(len(x)):
+                for key in x[0].keys():
+                    arr[i][key] = x[i][key]
+            return arr
+        return x
+
+    def _get_points_updates(self, batch_size: int) -> (npt.NDArray, npt.NDArray):
+        return self._to_array(self.gen.ask(batch_size)), self._to_array(self.gen.ask_updates())
+
+    def _convert_tell(self, x: npt.NDArray) -> list:
+        self.gen.tell(np_to_list_dicts(x))
+
+    def _loop_over_gen(self, tag, Work):
+        """Interact with ask/tell generator that *does not* contain a background thread"""
+        while tag not in [PERSIS_STOP, STOP_TAG]:
+            batch_size = self.batch or Work["libE_info"]["batch_size"]
+            points, updates = self._get_points_updates(batch_size)
+            if updates is not None and len(updates):  # returned "samples" and "updates". can combine if same dtype
+                H_out = np.append(points, updates)
+            else:
+                H_out = points
+            tag, Work, H_in = self.ps.send_recv(H_out)
```

Comment:
Dont we need a `np_to_list_dicts(H_out)` before this. 

Also, in the list of dictionaries data structure. If we convert a multi-dimensional x field, what type is the dictionary value supposed to be? Is it supposed to be a numpy array. I wonder this because the idea of the simple data structures was in part meant to avoid numpy.